### PR TITLE
Fix very slow regexp

### PR DESCRIPTION
**Pull request by caub**

### CODE REVIEW

**Comment by caub** [comment]
@formulahendry I did a very slight change to fix this 'infinite' running regex, (just `[^<>]*?` to `[^<>]?`)

some hints from: http://www.rexegg.com/regex-explosive-quantifiers.html

Could you merge it?

**Comment by caub** [comment]
@formulahendry Can you check it please?

**Comment by garygreen** [comment]
> I did a very slight change to fix this 'infinite' running regex, (just [^<>]*? to [^<>]?)

Yet the PR has hundreds of lines changed with formatting stuff. You might have more luck getting this merged if you changed a small amount of code... 😄

**Comment by caub** [comment]
ok back with some more shitty indentation

but I don't think @formulahendry is giving a lot of concern anyway

**Comment by caub** [comment]
single or double quotes? there is a mix of them in the sources

**Comment by caub** [comment]
I think I'll make a bot that gently ping @formulahendry every week, to hopefully fix this killing issue

File: package.json [hunk 1]
```diff
@@ -78,13 +78,30 @@
                 },
                 "auto-close-tag.excludedTags": {
                     "type": "array",
-                    "default": ["area", "base", "br", "col", "command", "embed", "hr", "img", "input", "keygen", "link", "meta", "param", "source", "track", "wbr"],
```
Comment (caub) [inline package.json]: `npm i` has reformatted it

File: package.json [hunk 2]
```diff
@@ -78,13 +78,30 @@
...
+                        "track",
+                        "wbr"
+                    ],
                     "description": "Set the tag list that would not be auto closed.",
                     "scope": "resource"
                 },
                 "auto-close-tag.SublimeText3Mode": {
                     "type": "boolean",
-                    "default": false,
+                    "default": true,
```
Comment (formulahendry) [inline package.json]: We'd not change the default setting if we could not make sure most of users want this.

File: src/getCloseTag.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
+export default function getCloseTag(text: string, excludedTags: string[]): string {
+    const regex = /<(\/?[a-zA-Z][a-zA-Z0-9:\-_.]*)(?:\s+[^<>]?[^\s/<>=]*)*\s?>/g;
```
Comment (formulahendry) [inline src/getCloseTag.ts]: `<abc key=value>` would not work correctly.

Comment (caub) [reply src/getCloseTag.ts]: ah yes indeed it's broken, need to add many more tests, thanks for noticing
```js
// with my change:
'lol <a href="/ok"></a><abc key="value"> </'.match(/<(\/?[a-zA-Z][a-zA-Z0-9:\-_.]*)(?:\s+[^<>]*?[^\s/<>=]+?)*?\s?>/g)
(3) ["<a href="/ok">", "</a>", "<abc key="value">"]
// original regex
'lol <a href="/ok"></a><abc key="value"> </'.match(/<(\/?[a-zA-Z][a-zA-Z0-9:\-_.]*)(?:\s+[^<>]?[^\s/<>=]*)*\s?>/g)
["</a>"]
```

File: src/getCloseTag.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
+export default function getCloseTag(text: string, excludedTags: string[]): string {
+    const regex = /<(\/?[a-zA-Z][a-zA-Z0-9:\-_.]*)(?:\s+[^<>]*?[^\s/<>=]*)*\s?>/g;
```
Comment (formulahendry) [inline src/getCloseTag.ts]: `'<abc =>` should not get close tag.

File: test/getCloseTag.test.ts [hunk 5]
```diff
@@ -0,0 +1,61 @@
+import { readFileSync } from "fs";
```
Comment (formulahendry) [inline test/getCloseTag.test.ts]: Need to enable test code running,  similar to  https://github.com/formulahendry/vscode-auto-rename-tag/blob/3141e782cadbb121b8c0fbf13683a88374e6a1b1/package.json#L54

Comment (formulahendry) [reply test/getCloseTag.test.ts]: Also need to follow the test structure: https://github.com/formulahendry/vscode-auto-close-tag/blob/master/test/extension.test.ts

File: src/getCloseTag.ts (hunk 4 above)
Comment (formulahendry) [reply src/getCloseTag.ts]: The biggest concern is that: also this may improve the performance, but might break several kinds of scenario,

Comment (formulahendry) [reply src/getCloseTag.ts]: `(?:\s+[^<>]*?[^\s/<>=]*)*` would make `[^\s/<>=]*` useless
`(?:\s+[^<>]*?[^\s/<>=]*)*` is similar to `(?:\s+[^<>]*?)*`

Comment (caub) [reply src/getCloseTag.ts]: Thanks for the advices  
what is `<abc =>` for? which languages please?

currently I hit again the very slow issue, I think I'll try a different approach than the big regex, and search backward with smaller regexes and a loop

Comment (formulahendry) [reply src/getCloseTag.ts]: arrow function shouldn't trigger auto-close: https://github.com/formulahendry/vscode-auto-close-tag/issues/5

Comment (formulahendry) [reply src/getCloseTag.ts]: search backward would be great idea! 👍

**Comment by caub** [comment]
I did that ^ finally

File: src/getCloseTag.ts (hunk 4 above)
Comment (caub) [reply src/getCloseTag.ts]: oh you meant `'<abc =>'` like in a string? (or a comment, ..). Honestly I don't think this extension can be perfect, with jsx expressions we'd need to parse them to be 100% correct, and it's not even possible when we are in an unfinished one. So we might need to accept false-positiives like `<foo><bar x=1 y={'<lol'} ></`  that gives `lol` currently

File: test/getCloseTag.test.ts (hunk 5 above)
Comment (caub) [reply test/getCloseTag.test.ts]: will try to add tests, (code-runner would need tests too, if I figure a good way, I'd update that project too)

**Comment by caub** [comment]
Ok, so I made a new approach by searching tags backward, it's not 100% robust, we can try to imrpove it more, like the current one anyway (jsx expressions can be very tricky), but no more performance issues at least

it's tested like the pointers you showed and run in CI

File: package.json (hunk 2 above)
Comment (joeysino) [reply package.json]: I want this!

File: src/getCloseTag.ts (hunk 4 above)
Comment (joeysino) [reply src/getCloseTag.ts]: I don't mind if a few rare edge cases are broken, if that means the performance issues are fixed. Performance a big concern for this extension.

(I just wish we could add a timeout to regexps so they will fail if they take too long!)

**Comment by garethedwards-tass** [comment]
This has helped me a bit, thank you, but also your code wasn't catering for excludedTags, and I think this does the job.

`const TAG_RE = /<(\/?[a-zA-Z][a-zA-Z0-9:_.-]*)(?![\s\S]*<\/?[a-zA-Z])/;

function getCloseTag(text: string, excludedTags: string[]): string {
    const s = text[text.length - 1] === '/' && text[text.length - 2] === '<' ? text.slice(0, -2) : text[text.length - 1] === '<' ? text.slice(0, -1) : text;
    let m = s.match(TAG_RE);
    // while we catch a closing tag, we jump directly to the matching opening tag
    while (m && ( m[1][0] === '/' || excludedTags.indexOf(m[1].toLowerCase()) !== -1 )) {
        const s2 = s.slice(0, m.index);
        if ( m[1][0] === '/' ) {
            // Already Closed Tags
            const m2 = s2.match(RegExp(`<${m[1].slice(1)}.*$`, 'm'));
            if (!m2) return '';
            m = s.slice(0, m2.index).match(TAG_RE);
        } else {
            // Excluded Tags
            m = s.slice(0, m.index).match(TAG_RE);
        }
    }
    if (!m) return null;
    return (text[text.length - 1] === '/' && text[text.length - 2] === '<' ? m[1] : text[text.length - 1] === '<' ? '/' + m[1] : '</' + m[1]) + '>';
}
`

**Comment by caub** [comment]
Sorry guys I can't maintain this PR anymore, please go ahead with a new PR

**Comment by garethedwards-tass** [comment]
@caub curious, are you using different tools or just not this extension anymore ? I've built in some of this functionality into our internal tools to avoid devs having to use this extension.

**Comment by caub** [comment]
I went back to sublime-text

**Comment by garethedwards-tass** [comment]
@caub we took this route a few times in the past, curious what languages you're working with? we're writing CFML and I've made some headway using your code, its not perfect but it works ok.

**Comment by caub** [comment]
@garethedwards-tass you can definitely make progress, try to make you own PR and the maintainer may merge it

I'm only working on JS (node+react), so sublimetext support things decently

**Pull request closed without merging**